lib/Frontend/Frontend.cpp

```diff
@@ -440,141 +549,150 @@ void CompilerInstance::performSema() {
+      performTypeChecking(MainFile, PersistentState.getTopLevelContext(),
+                          TypeCheckOptions, CurTUElem,
+                          options.WarnLongFunctionBodies,
+                          options.WarnLongExpressionTypeChecking,
+                          options.SolverExpressionTimeThreshold);
+    }
+    CurTUElem = MainFile.Decls.size();
+  } while (!Done);
+
+  Diags.setSuppressWarnings(DidSuppressWarnings);
+
+  if (mainIsPrimary && !Context->hadError() &&
+      Invocation.getFrontendOptions().PCMacro) {
+    performPCMacro(MainFile, PersistentState.getTopLevelContext());
+  }
+
+  // Playground transform knows to look out for PCMacro's changes and not
+  // to playground log them.
+  if (mainIsPrimary && !Context->hadError() &&
+      Invocation.getFrontendOptions().PlaygroundTransform)
+    performPlaygroundTransform(
+        MainFile, Invocation.getFrontendOptions().PlaygroundHighPerformance);
+  if (!mainIsPrimary) {
+    performNameBinding(MainFile);
+  }
+}
+
+void CompilerInstance::typeCheckEveryFileInMainModule(
+                                                      PersistentParserState &PersistentState,
+                                                      const OptionSet<TypeCheckingFlags> TypeCheckOptions) {
+  const auto &options = Invocation.getFrontendOptions();
+  for (auto File : MainModule->getFiles()) {
+    if (auto SF = dyn_cast<SourceFile>(File)) {
+      performTypeChecking(*SF, PersistentState.getTopLevelContext(),
+                          TypeCheckOptions, /*curElem*/ 0,
+                          options.WarnLongFunctionBodies,
+                          options.WarnLongExpressionTypeChecking,
+                          options.SolverExpressionTimeThreshold);
+    }
+  }
+}
+void CompilerInstance::typeCheckThePrimaryFile(
+                                               PersistentParserState &PersistentState,
+                                               const OptionSet<TypeCheckingFlags> TypeCheckOptions) {
+  assert (PrimarySourceFile != nullptr  &&  "No primary file?");
+  const auto &options = Invocation.getFrontendOptions();
+  performTypeChecking(*PrimarySourceFile, PersistentState.getTopLevelContext(),
+                      TypeCheckOptions, /*curElem*/ 0,
+                      options.WarnLongFunctionBodies,
+                      options.WarnLongExpressionTypeChecking,
+                      options.SolverExpressionTimeThreshold);
+}
+
+
+
+void CompilerInstance::typeCheckMainModule(
+    OptionSet<TypeCheckingFlags> TypeCheckOptions) {
+  SharedTimer timer("performSema-checkTypesWhileParsingMain-typeCheckMainModule");
   if (TypeCheckOptions & TypeCheckingFlags::DelayWholeModuleChecking) {
-    for (auto File : MainModule->getFiles())
-      if (auto SF = dyn_cast<SourceFile>(File))
-        performWholeModuleTypeChecking(*SF);
+    performWholeModuleTypeCheckingOnMainModule();
   }
+  finishTypeCheckingMainModule();
```

Comment:
I don't think these two operations are worth grouping any more than the rest of what's now 'checkTypesWhileParsingMain'.

Reply:
Thanks for your patience: take a look at what it is now, and let me know what you'd like. It's changed a bit since this comment.